mdata/cache/ccache_metric.go

```diff
@@ -69,6 +69,66 @@ func (mc *CCacheMetric) Del(ts uint32) int {
 	return len(mc.chunks)
 }
 
+func (mc *CCacheMetric) AddSlice(prev uint32, itergens []chunk.IterGen) {
+	if len(itergens) == 0 {
+		return
+	}
+
+	mc.Lock()
+	defer mc.Unlock()
+
+	ts := itergens[0].Ts
+
+	// if previous chunk has not been passed we try to be smart and figure it out.
+	// this is common in a scenario where a metric continuously gets queried
+	// for a range that starts less than one chunkspan before now().
+	if prev == 0 {
+		res, ok := mc.seekDesc(ts - 1)
+		if ok {
+			prev = res
+		}
+	}
+
+	for _, itergen := range itergens {
+		ts = itergen.Ts
+
+		if _, ok := mc.chunks[ts]; ok {
+			// chunk is already present. no need to error on that, just ignore it
+			continue
+		}
+
+		mc.chunks[ts] = &CCacheChunk{
+			Ts:    ts,
+			Prev:  0,
+			Next:  0,
+			Itgen: itergen,
+		}
+
+		// if the previous chunk is cached, link in both directions
+		if _, ok := mc.chunks[prev]; ok {
+			mc.chunks[prev].Next = ts
+			mc.chunks[ts].Prev = prev
+		}
+		prev = ts
```

Comment:
this is buggy, because higher up we continue if a chunk exists, but we won't update the prev value, and the next chunk we'll need to insert wil set up the links so that pre-existing chunks will be skipped.